[JENKINS-54250] Disable restarting stages in Jenkinsfile optionally

pavlovic-ivan opened the pull request:
# Description

See [JENKINS-54250](https://issues.jenkins-ci.org/browse/JENKINS-54250). 

### Instructions
**Case 1**
- configure pipeline using JenkinsUI
- add new option
```
options {
  disableRestartFromStage()
}
```
- save the changes
- schedule a new build with JenkinsUI or Blueocean
- open the latest build with Blueocean
- the "Restart from stage" button should not be visible

**Case 2**
- configure pipeline using JenkinsUI
- remove new option
```
options {
}
```
- save the changes
- schedule a new build with JenkinsUI or Blueocean
- open the latest build with Blueocean
- the "Restart from stage" button is visible again

### NOTE: This PR depends on this [one](https://github.com/jenkinsci/pipeline-model-definition-plugin/pull/560). Until that one is merged, this one can not be merged as well

# Submitter checklist
- [x] Link to JIRA ticket in description, if appropriate.
- [x] Change is code complete and matches issue description
- [x] Appropriate unit or acceptance tests or explanation to why this change has no tests
- [x] Reviewer's manual test instructions provided in PR description. See Reviewer's first task below.

# Reviewer checklist
- [ ] Run the changes and verified the change matches the issue description
- [ ] Reviewed the code
- [ ] Verified that the appropriate tests have been written or valid explanation given

pavlovic-ivan commented:
Hi @olamy. This would be the PR for the fixes discussed in another [PR](https://github.com/jenkinsci/pipeline-model-definition-plugin/pull/560) as well. Your help and feedback is much appreciated! 🙏

pavlovic-ivan commented:
@olamy with i was 3 commits behind master. After rebasing i see that `powermock` is removed. What is preferred approach now?

olamy commented:
> @olamy with i was 3 commits behind master. After syncing i see that `powermock` is removed. What is preferred approach now?

yup this has been removed and not working well with Java 17.
you must use plain mockito now.

pavlovic-ivan commented:
Hi @olamy. Thanks for all the help. I've moved away from `powermock` and used `mockito`. New tests are green when i run them locally. I see the pipeline has failed, but the build was a success.

Looking forward to your feedback 🚀

olamy commented:
@pavlovic-ivan thanks for that!
I reproduce the failure locally (using java11)
```
[ERROR] Tests run: 6, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 4.462 s <<< FAILURE! - in io.jenkins.blueocean.rest.impl.pipeline.PipelineNodeImplTest
[ERROR] io.jenkins.blueocean.rest.impl.pipeline.PipelineNodeImplTest.unnecessary Mockito stubbings  Time elapsed: 0.027 s  <<< ERROR!
org.mockito.exceptions.misusing.UnnecessaryStubbingException: 

Unnecessary stubbings detected in test class: PipelineNodeImplTest
Clean & maintainable test code requires zero unnecessary code.
Following stubbings are unnecessary (click to navigate to relevant line of code):
  1. -> at io.jenkins.blueocean.rest.impl.pipeline.PipelineNodeImplTest.getRun_isRestartable(PipelineNodeImplTest.java:94)
  2. -> at io.jenkins.blueocean.rest.impl.pipeline.PipelineNodeImplTest.getRun_isNotRestartable(PipelineNodeImplTest.java:133)
  3. -> at io.jenkins.blueocean.rest.impl.pipeline.PipelineNodeImplTest.getRun_isNotRestartable(PipelineNodeImplTest.java:139)
  4. -> at io.jenkins.blueocean.rest.impl.pipeline.PipelineNodeImplTest.getRun_isNotRestartable(PipelineNodeImplTest.java:140)
Please remove unnecessary stubbings or use 'lenient' strictness. More info: javadoc for UnnecessaryStubbingException class.
	at org.mockito.internal.runners.StrictRunner.run(StrictRunner.java:53)
	at org.mockito.junit.MockitoJUnitRunner.run(MockitoJUnitRunner.java:163)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:365)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:273)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:238)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:159)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:377)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:138)
	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:465)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:451)
```

pavlovic-ivan commented:
> @pavlovic-ivan thanks for that! I reproduce the failure locally (using java11)
> 
> ```
> [ERROR] Tests run: 6, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 4.462 s <<< FAILURE! - in io.jenkins.blueocean.rest.impl.pipeline.PipelineNodeImplTest
> [ERROR] io.jenkins.blueocean.rest.impl.pipeline.PipelineNodeImplTest.unnecessary Mockito stubbings  Time elapsed: 0.027 s  <<< ERROR!
> org.mockito.exceptions.misusing.UnnecessaryStubbingException: 
> 
> Unnecessary stubbings detected in test class: PipelineNodeImplTest
> Clean & maintainable test code requires zero unnecessary code.
> Following stubbings are unnecessary (click to navigate to relevant line of code):
>   1. -> at io.jenkins.blueocean.rest.impl.pipeline.PipelineNodeImplTest.getRun_isRestartable(PipelineNodeImplTest.java:94)
>   2. -> at io.jenkins.blueocean.rest.impl.pipeline.PipelineNodeImplTest.getRun_isNotRestartable(PipelineNodeImplTest.java:133)
>   3. -> at io.jenkins.blueocean.rest.impl.pipeline.PipelineNodeImplTest.getRun_isNotRestartable(PipelineNodeImplTest.java:139)
>   4. -> at io.jenkins.blueocean.rest.impl.pipeline.PipelineNodeImplTest.getRun_isNotRestartable(PipelineNodeImplTest.java:140)
> Please remove unnecessary stubbings or use 'lenient' strictness. More info: javadoc for UnnecessaryStubbingException class.
> 	at org.mockito.internal.runners.StrictRunner.run(StrictRunner.java:53)
> 	at org.mockito.junit.MockitoJUnitRunner.run(MockitoJUnitRunner.java:163)
> 	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:365)
> 	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:273)
> 	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:238)
> 	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:159)
> 	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:377)
> 	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:138)
> 	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:465)
> 	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:451)
> ```

Thanks for the feedback. I'll take further look into it!

pavlovic-ivan commented:
@olamy i've fixed the tests that i've added. however, i am not sure why the `GitScmTest` tests are failing. I don't think i have something to do with them, and they are green with java 8

olamy commented:
@pavlovic-ivan your PR still says `DO NOT MERGE`?

pavlovic-ivan commented:
@olamy yes because it is waiting for this [PR](https://github.com/jenkinsci/pipeline-model-definition-plugin/pull/560) to be merged

pavlovic-ivan commented on blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/PipelineNodeImpl.java:
```diff
@@ -202,7 +202,16 @@ private WorkflowRun getRun() {
     @Override
     public boolean isRestartable() {
-        return restartable;
+        RestartDeclarativePipelineAction restartDeclarativePipelineAction =
```
ah, i see. looks good, i'll revert my change

pavlovic-ivan replied:
looks like i need to update the tests as well, as Mockito was mocking what was inside the `isRestartable` method

pavlovic-ivan commented:
Hi @olamy . Do i need to bump the version of the `pipeline-model-definition-plugin`, or is this done automatically?

olamy commented:
> Hi @olamy . Do i need to bump the version of the `pipeline-model-definition-plugin` in Blueocean dependencies, or is this done automatically?

if your PR depends on it. just do it with your PR to have a consistent change

pavlovic-ivan commented:
It's needed because of the tests, because the new option is added in the `pipeline-model-definition-plugin`. I suspect the checks will continue to fail, as tests that i've added will fail as well. When it comes to increasing the version, i can't seem to find where it is done. Is `hpi` plugin responsible for this?

pavlovic-ivan commented:
Hi @olamy . Sorry for bothering you. Do you maybe have any feedback for this PR?

olamy commented:
> Hi @olamy . Sorry for bothering you. Do you maybe have any feedback for this PR? Should i include `@jenkinsci/blueocean-plugin-developers` as well?

Sounds good but you might need to get the build pass ;) 
There are some upper bound rules issues to fix

pavlovic-ivan commented:
All right. Enforcer doesn't complain with the master branch, but it does complain when i bump the version of the dependency as i did:
```
<dependency>
    <groupId>org.jenkinsci.plugins</groupId>
    <artifactId>pipeline-model-definition</artifactId>
    <version>2.2125.vddb_a_44a_d605e</version>
</dependency>
```

i am not sure what i can do here, as i think all of this (version management) should've been done within other dependency manifests when plugin was released, coming from either a parent dependency, BOM, or something else. Please correct me if i am wrong, but i am not sure what is the best course of action here. What am i missing?

pavlovic-ivan commented:
Hi @olamy . Any thoughts on the last comment i've posted? Thanks in advance 🙏
